src/training.py

```diff
@@ -0,0 +1,79 @@
+import torch
+import wandb
+from tqdm import tqdm
+
+from src.diffusion.utils import configure_schedule
+from src.diffusion.utils import flat_mean
+from src.diffusion.utils import get_diffusion_variables
+from src.metrics import calculate_batch_ce
+from src.metrics import calculate_ce
+
+
+def train_model(
+    model,
+    train_dataloader,
+    val_dataloader,
+    schedule: str,
+    num_steps: int,
+    device: str,
+    project_name: str = "didi",
+    logging_step: int = 100,
+    step_freq: int = 10,
+):
+    wandb.init(project=project_name)
+
+    optimizer = torch.optim.AdamW(model.parameters(), lr=0.0001)
+
+    alphas_cumprod_prev = configure_schedule(model.diffusion_steps, schedule).to(device)
+
+    model.train()
+    with tqdm(total=num_steps) as pbar:
+        while num_steps:
+            # b_context: [batch size, context seq len], b_gt: [batch size, target seq len]
+            for b_context, b_gt in train_dataloader:
+                context = b_context.to(device)
+                gt = b_gt.to(device)
+
+                emb = model.emb(gt)
+
+                x_0, x_t, t = get_diffusion_variables(model.diffusion_steps, emb, alphas_cumprod_prev, device)
+
+                x_0_hat = model(
+                    encoder_input_ids=context,
+                    decoder_inputs_embeds=x_t,
+                    time_ids=t,
+                )
+
+                optimizer.zero_grad()
+
+                pad_mask = gt == model.emb.padding_idx
+                ce = calculate_batch_ce(model, x_0, gt, pad_mask)
```

Comment:
Are you sure that we should calculate loss using `x_0`? I looked into the Diffusion-LM code, the use model output for this (`x_0_hat` in our case) -- [code](https://github.com/XiangLi1999/Diffusion-LM/blob/main/improved-diffusion/improved_diffusion/gaussian_diffusion.py#L1688)

Reply:
Hmm, but other implementation uses `x_0` -- [code](https://github.com/maximzubkov/minimal-text-diffusion/blob/main/src/modeling/diffusion/gaussian_diffusion.py#L295)